### PR TITLE
cmd/kubelet: Introduce --bootstrap-pod-label flag

### PR DESCRIPTION
**Pull request by derekparker**
This patch introduces a new flag to the kubelet, the
--bootstrap-pod-label flag. This allows the user to provide a
`key=value` label to the kubelet to indicate that the kubelet should run
until seeing a pod with a matching label. Once a pod with the matching
label is found, the kubelet will run it and then exit.

This is useful for bootstrapping a self-hosted cluster. Once the
"bootstrap" kubelet dies, it could effectively bring up a new, managed
kubelet, and continue the initial self-hosted cluster bootstrapping.

### CODE REVIEW

**Comment by derekparker** [comment]
cc: @aaronlevy @mikedanese (feel free to cc in anyone else interested in self-hosted).

**Comment by vishh** [comment]
I'd prefer discussing a solution for kubelet bootstrapping through a proposal instead of a PR.

File: cmd/kubelet/app/options/options.go [hunk 1]
```diff
@@ -183,6 +185,8 @@ func (s *KubeletServer) AddFlags(fs *pflag.FlagSet) {
 	fs.DurationVar(&s.NodeStatusUpdateFrequency.Duration, "node-status-update-frequency", s.NodeStatusUpdateFrequency.Duration, "Specifies how often kubelet posts node status to master. Note: be cautious when changing the constant, it must work with nodeMonitorGracePeriod in nodecontroller. Default: 10s")
 	bindableNodeLabels := util.ConfigurationMap(s.NodeLabels)
 	fs.Var(&bindableNodeLabels, "node-labels", "<Warning: Alpha feature> Labels to add when registering the node in the cluster.  Labels must be key=value pairs separated by ','.")
+	bindableBootstrapPodLabel := util.ConfigurationMap(s.BootstrapPodLabel)
+	fs.Var(&bindableBootstrapPodLabel, "bootstrap-pod-label", "Label used to filter pods when looking for bootstrap pod, must be key=value pair. Kubelet will run until seeing a pod with this label, after which it will run that pod and exit.")
```
Comment (mikedanese) [inline cmd/kubelet/app/options/options.go]: --runonce-pod-label-selector might be a more descriptive name

**Comment by luxas** [comment]
Please cc me on self-hosting conversations

**Comment by mikedanese** [comment]
@dchen1107

**Comment by derekparker** [comment]
@vishh Completely understandable.

We have had some discussion of this approach internally at CoreOS as well as via the sig-high-availability group. We decided to implement and build a self-hosted POC around this approach (which I could demo at the next sig-ha). We had also considered exploiting the existing runonce behavior with a few code modifications, but came to the conclusion that an explicit bootstrap label combined with runonce semantics gave us the deterministic bootstrap / pivot we were looking for.

I'm happy to open a dialogue around this PR, and further enumerate other approaches and the pros/cons.

**Comment by vishh** [comment]
@derekparker: Got it. A proposal with alternatives considered, along with pros & cons will be very helpful, not just for this PR, but to make sure the entire community is on board.

**Comment by mikedanese** [comment]
The two required modifications for the approach @coreos is testing out seem small enough and generally useful (this PR and one other PR that decouples runonce mode from standalone mode) and would not conflict with or preclude us from going with other approaches. I would like these features. We can also pilot this flag as an --alpha flag.

I don't think we need to settle on a single approach to bootstrapping either. We are in the contentious realm of deployment where everyone has opposing opinions and no one will budge. We can allow for flexibility here like we allow for flexibility in other areas of kubernetes deployment. If we do want to decide on a single approach a design doc should go through the proposal process, but it's a ton of overhead for small features that are not enabled by default. It will slow down experimentation.

**Comment by mikedanese** [comment]
Why should this feature be specific to runonce? I haven't decided one way or the other. Are static pods subject to canAdmitPods?

**Comment by vishh** [comment]
Over time, maintaining features that are optionally used will result in
significant maintenance burden not only for kubelet but for any software.
Each optional feature we add will usually be very simple. But together,
they lead to maintenance burden.
Ideally, every feature we add should be well tested. I'd really like to see
the alternatives considered before we start merging a bunch of features
that are trying to provide the same functionality.
I don't mean to slow down the progress here, and I'm willing to guarantee
quick reviews.

One more alternative to consider might be that of working against a
separate bootstrapping branch to test out various alternatives before
pushing one or more of the solutions to master.

**Comment by mikedanese** [comment]
@derekparker @aaronlevy can you morph your doc into a design proposal for review by a larger audience with a blurb that explains why you chose this approach over others?

**Comment by derekparker** [comment]
@mikedanese @vishh absolutely, we can get started on putting that together.

**Comment by derekparker** [comment]
After speaking with @mikedanese and others internally, I am going to close out this PR in favor of a similar but less intrusive (to the kubelet) approach.

**Comment by mikedanese** [comment]
I think we can use taints and tolerations once that is implemented.

**Pull request closed without merging**